src/repository/product/product.repository.ts

```diff
@@ -0,0 +1,90 @@
+import { DataSource, Repository } from 'typeorm';
+import {
+  ConflictException,
+  Injectable,
+  InternalServerErrorException,
+  Logger
+} from '@nestjs/common';
+import { ErrorCodes } from '@enums/error_codes.enum';
+import { Product } from '@entities/product';
+import { CreateProductDto } from '@dtos/product';
+
+@Injectable()
+export class ProductRepository extends Repository<Product> {
+  private logger: Logger = new Logger('ProductRepository', { timestamp: true });
+  constructor(private dataSource: DataSource) {
+    super(Product, dataSource.createEntityManager());
+  }
+
+  async createProduct(createProductDto: CreateProductDto): Promise<Product> {
+    const { name, description, price, stock } = createProductDto;
+
+    const product = this.create({
+      name,
+      description,
+      price,
+      stock
+    });
+
+    try {
+      await this.save(product);
+
+      this.logger.verbose(`Created Product ${product.name}`);
+
+      return product;
+    } catch (error) {
+      this.logger.error(
+        `Error creating Product. Product name: ${name}`,
+        error.stack
+      );
+      if (error.code === ErrorCodes.DB_DUPLICATE_VALUE) {
+        throw new ConflictException('Product name already in use');
+      } else {
+        throw new InternalServerErrorException();
+      }
+    }
+  }
+
+  async findByName(name: string): Promise<Product> {
+    try {
+      const product = await this.findOneBy({ name });
```

Comment:
here you could just return the Repository promise instead of declare a `product` variable